RFD 208: Add ProxyWindowsDesktopSession API to Teleport Proxy Service (#54070)
gravitational/teleport: rfd/0208-proxy-desktop-session.md

## Details

In the Web UI, the TDP stream is transmitted over a cookie-authenticated WebSocket. 

However, Teleport Connect (which uses tsh under the hood) does not rely on cookies for authentication—instead, it uses client certificates.
To support this scenario, a new gRPC-based API called `ProxyDesktopSession` will be added to the `TransportService` in the Teleport Proxy. 

### Architecture

```
            gRPC bidi stream   
+-----+   (ProxyDesktopSession)    +----------------+      TCP connection     +-------------------------+
| tsh | <=======================>  |  Proxy Service | <=====================> | Windows Desktop Service |
+-----+                            +----------------+                         +-------------------------+
        <===================================================================>
                              TLS connection (end-to-end)
```

`ProxyDesktopSession` won't transmit TDP messages directly.
Instead, it will tunnel a TLS connection over gRPC, which will then carry the TDP messages.

> This design simplifies per-session MFA by allowing the client to present an MFA-verified certificate over a single TLS connection that terminates at the Windows Desktop Service.

Reviewer: How does the client know that it needs to obtain a MFA-verified cert?

Author: I wanted to generate the cert with `TeleportClient.IssueUserCertsWithMFA`, which calls `IsMFARequired` under the hood.

```go
cert, err := clusterClient.IssueUserCertsWithMFA(ctx, client.ReissueParams{
	RouteToCluster: clusterClient.SiteName,
	TTL:            clusterClient.KeyTTL,
	RouteToDesktop: proto.RouteToWindowsDesktop{
		WindowsDesktop: desktopName,
		Login:          login,
	},
})
```

Reviewer: Oh, we need a dedicated per-desktop cert anyway, I see.

Author: Yes, I didn't mention it in the RFD, but in tsh we will need to add support for windows desktops certs.

2. The Teleport Proxy uses the provided `dial_target` to dial the Desktop Service, following the logic in a Web UI handler.
   * The Proxy queries the cluster's desktops looking for the ones that match the `desktop_name` in `dial_target`.
   * The returned desktops are then mapped to desktop services reporting them.
   * The Proxy attempts to connect to these services in a randomized order until one succeeds.
3. Data tunneling starts.
   * Once a Desktop Service is successfully reached, the Proxy begins forwarding raw data between tsh and the Desktop Service.
   * Meanwhile, tsh initiates a TLS connection over the gRPC stream using its MFA-verified client certificate.
   * The TLS connection is established when the Desktop Service accepts the handshake.

Note that Proxy won't enforce RBAC checks. It will attempt to connect to the Windows Desktop Service even if the user doesn’t have access to the target desktop.
The Desktop Service will enforce access control and return an error if the user isn’t allowed.

### Drawbacks and considered alternatives

Tunneling a TLS connection over a gRPC stream introduces some overhead due to the added TLS metadata in each message. 

> This could impact performance slightly, although manual testing did not show any noticeable latency.

Reviewer: Something has to run the TLS connection to the service tho, and my immediate thought is that it's better to spread the cost of that to the clients rather than have the proxy do it for everyone.

A potential alternative design would terminate the client's TLS connection in the Proxy, then open a separate TLS connection to the Desktop Service. 
This would require the Proxy to detect whether per-session MFA is needed, issue an MFA challenge to the client and verify the response.
However, this approach is not currently feasible, since Proxy has no straightforward way to determine whether per-session MFA is required 
for a given user—it doesn’t interact with the Auth Service using a user's identity.

### gRPC Definitions

#### Service definition

```protobuf
// Bidirectional stream for proxying desktop sessions.
rpc ProxyDesktopSession(stream ProxyDesktopSessionRequest) returns (stream ProxyDesktopSessionResponse);
```

#### Message definitions

```protobuf
// Request message for a proxied desktop session.
message ProxyDesktopSessionRequest {

>   // Target cluster and desktop. Must be sent as the first message.

Reviewer: ```suggestion
  // Target cluster and desktop. Must be set in the first message and unset in subsequent messages.
```

  TargetDesktop dial_target = 1;

>    // Payload from Teleport Desktop Protocol. 
>    DesktopSessionMessage message = 2;

Reviewer: Unless there's some out of band data that we have to send, try to keep these messages as light as possible - the cost of handling this grpc stream in the proxy is basically only grpc/protobuf costs because all we do is copy the data and push it somewhere else.

```suggestion
    // A chunk of data from the connection. Can be nonempty even in the first message, but it's also legal for it to be empty.
    bytes data = 2;
```

Reviewer: (also for symmetry with the response maybe put the data on field 1 and the dial req on 2? I can imagine some specialized grpc codec that only deals with a single bytes field and doing it just for field 1 would be a tiny bit easier I think)

Author: Right, we don't really need to wrap the data in another message.

}

// Response message for a proxied desktop session.
message ProxyDesktopSessionResponse {

>    // Payload from Teleport Desktop Protocol.
>    DesktopSessionMessage message = 1;

Reviewer: ```suggestion
    // A chunk of data from the connection. Can be empty (for example, to send a message signaling a successful connection even if there's no data available in the connection).
    bytes data = 1;
```

> }

Reviewer: Do we need any info about the cluster like we do in SSH, like whether or not we're FIPS?

Author: AFAIK we don't need such info.

// Identifies the destination desktop within a specific cluster.
message TargetDesktop {
   // Operating system of the target desktop.

>    DesktopOS desktop_os = 1;

Author: I think it's worth adding the target operating system. The name `ProxyDesktopSession` doesn’t indicate which OS it's for, but the current implementation will only support Windows. Since Linux desktop access is on the roadmap, including the OS in the name should make the API more future-proof.

Reviewer: Are we even sure that the behavior at the proxy will be the same for linux desktops - i.e. passing along a bytestream given a name in the form of a string and a cluster name?

Reviewer: There's also no real cost in supporting multiple RPCs that do the same thing, so we could just call the rpc `ProxyWindowsDesktopSession` and leave the problem for a later time.

Author: This is a good idea, I renamed the RPC to `ProxyWindowsDesktopSession`. Thanks for the suggestion!
